### PR TITLE
Junior Rollercoaster Track Paint

### PR DESCRIPTION
**Pull request by duncanspumpkin**
WIP

I plan to implement all of the junior rollercoaster track paint functions and then look into how best to make this modular for the rest of the rollercoaster tracks. At present 5 pieces have been implemented. (<250 to go! {Junior Rollercoasters don't have every type of track})

jrcprogress1

### CODE REVIEW

File: src/ride/track_paint.c [hunk 1]
```diff
@@ -1139,3 +1144,321 @@ TRACK_PAINT_FUNCTION get_track_paint_function_facility(int trackType, int direct
...
+	sint8 offsetX = direction & 1 ? 6 : 0;
+	//cl
+	sint8 offsetY = direction & 1 ? 0 : 6;
+
+	sub_98196C(image_id, offsetX, offsetY, lengthX, lengthY, 1, height, get_current_rotation());
+
+	if ((RCT2_GLOBAL(0x009DE56A, sint16)& (1 << 5) && !(RCT2_GLOBAL(0x009DE56E, sint16)& (1 << 5))) || 
+		!(RCT2_GLOBAL(0x009DE56A, sint16)& (1 << 5) && (RCT2_GLOBAL(0x009DE56E, sint16)& (1 << 5)))) {
+		int edi = direction & 1 ? 2 : 1;
+		RCT2_CALLPROC_X(0x00663105, 0, 4, 0, height, 20, edi, RCT2_GLOBAL(0x00F4419C, uint32));
```
Comment (marijnvdwerf) [inline src/ride/track_paint.c]: This should be `metal_a_wooden_a_supports_paint_setup`. `ebx` is the desired segment to put the pole on. 
```
    0
  5   6
1   4   2
  7   8
    3
```

File: src/ride/track_paint.c [hunk 2]
```diff
@@ -1139,3 +1144,321 @@ TRACK_PAINT_FUNCTION get_track_paint_function_facility(int trackType, int direct
...
+/* 0x008AAA0C */
+TRACK_PAINT_FUNCTION get_track_paint_function_junior_rc(int trackType, int direction) {
+	switch (trackType) {
+	case 0:
+		return junior_rc_flat_paint_setup;
+	case 1:
+		return junior_rc_end_station_paint_setup;
+	}
+	return NULL;
+}
```
Comment (LinusU) [inline src/ride/track_paint.c]: No newline at end of file...

File: src/ride/track_paint.c (hunk 1 above)
Comment (duncanspumpkin) [inline src/ride/track_paint.c]: Yeah i had to change it to not use the metal_a_ function call as i realised ebx at present is not passed into that function.

Comment (marijnvdwerf) [inline src/ride/track_paint.c]: Is the function used anywhere else yet?

File: src/ride/coaster/junior_roller_coaster.c [hunk 3]
```diff
@@ -13,3 +13,744 @@
...
+	if (!(RCT2_GLOBAL(0x00F441A0, uint32) & (1 << 29))) {
+		imageId &= 0x7FFFF;
+	}
+
+	RCT2_GLOBAL(0x00F441E8, uint32) = imageId;
+	RCT2_GLOBAL(0x00F441E4, uint32) = imageId;
+
+	bool hasFence;
+	bool hasGreenLight = (bool) (mapElement->properties.track.sequence & 0x80);
+	bool isBreaked = (bool)(mapElement->flags & MAP_ELEMENT_FLAG_BLOCK_BREAK_CLOSED);
```
Comment (Gymnasiast) [inline src/ride/coaster/junior_roller_coaster.c]: Wrong spelling. It's `brake`.

**Comment by Gymnasiast** [comment]
You still need to correct the usage of 'break' instead of 'brake'.

**Comment by duncanspumpkin** [comment]
Yeah will get on that. I need to redo a huge load of it to match up with everything else. again

**Comment by duncanspumpkin** [comment]
28/82 complete

File: src/ride/track_paint.c [hunk 4]
```diff
@@ -551,6 +551,64 @@ void track_paint_util_draw_pier(rct_ride * ride, const rct_ride_entrance_definit
...
+		{32, 16},
+		{32, 20},
+	}
+};
+
+static const sint8 right_helix_up_small_tiles_sprite_map[] = {0, -1, 1, 2, 3, -1, 4, 5};
+void track_paint_util_right_helix_up_small_tiles_paint(const sint8 thickness[2], sint16 height, int direction, uint8 trackSequence, uint32 colourFlags, const uint32 sprites[4][6][2], const rct_xy16 offsets[4][6][2], const rct_xy16 boundsLengths[4][6][2], const rct_xy16 boundsOffsets[4][6][2], uint8 rotation)
+{
+	int index = right_helix_up_small_tiles_sprite_map[trackSequence];
+	if (index < 0) {
```
Comment (marijnvdwerf) [inline src/ride/track_paint.c]: Shouldn't this be `== 0`?

File: src/ride/coaster/junior_roller_coaster.c [hunk 5]
```diff
@@ -101,7 +101,7 @@ void paint_junior_rc_station_track(uint8 rideIndex, uint8 trackSequence, uint8 d
 	RCT2_GLOBAL(0x00F441E8, uint32) = imageId;
```
Comment (janisozaur) [inline src/ride/coaster/junior_roller_coaster.c]: (regarding commit message): `splleing` :P

Comment (SijmenSchoon) [inline src/ride/coaster/junior_roller_coaster.c]: @janisozaur Somehow I think that's not accidental :p

**Comment by duncanspumpkin** [comment]
82/82 complete :)

**Comment by Manuel-K** [comment]
How many track types (that the Junior Rollercoaster doesn't have) are in the game?

**Comment by duncanspumpkin** [comment]
There are 255 track types. Some of them are completely useless for coasters I.e. golf course holes and such. So Jr RC with 82 is a sizable chunk of them.

**Comment by Manuel-K** [comment]
@duncanspumpkin I've just noticed that I forgot to thank you for your answer. So: Thank you!

Keep up the good work.

**Pull request merged**